main/src/ca/uwaterloo/flix/language/phase/unification/HybridFormula.scala

```diff
@@ -0,0 +1,190 @@
+    case bddForm: BddFormula => bddForm
+  }
+
+  def convertBoolFormToBddForm(f: BoolFormula): BddFormula = f match {
+    case BoolFormula.True => bddAlg.mkTrue
+    case BoolFormula.False => bddAlg.mkFalse
+    case BoolFormula.Var(id) => bddAlg.mkVar(id)
+    case BoolFormula.And(f1, f2) =>
+      bddAlg.mkAnd(convertBoolFormToBddForm(f1), convertBoolFormToBddForm(f2))
+    case BoolFormula.Or(f1, f2) =>
+      bddAlg.mkOr(convertBoolFormToBddForm(f1), convertBoolFormToBddForm(f2))
+    case BoolFormula.Not(f1) => bddAlg.mkNot(convertBoolFormToBddForm(f1))
+  }
+
+  def convertToBoolFormula(f: Formula): BoolFormula = f match {
+    case boolForm: BoolFormula => boolForm
+    case bddForm: BddFormula => convertBddFormToBoolForm(bddForm.dd, boolAlg.mkTrue)
+  }
+
+  def convertBddFormToBoolForm(dd: DD, acc: BoolFormula): BoolFormula = {
+    if (dd.isLeaf()) {
+      return if (dd.isTrue()) acc else boolAlg.mkFalse
+    }
+
+    val currentVar = dd.getVariable()
+    val varForm = boolAlg.mkVar(currentVar)
+
+    val lowForm = boolAlg.mkAnd(acc, boolAlg.mkNot(varForm))
+    val lowRes = convertBddFormToBoolForm(dd.getLow(), lowForm)
+    val highForm = boolAlg.mkAnd(acc, varForm)
+    val highRes = convertBddFormToBoolForm(dd.getHigh(), highForm)
+
+    (lowRes, highRes) match {
+      case (BoolFormula.False, BoolFormula.False) => BoolFormula.False
+      case (BoolFormula.False, _) => highRes
+      case (_, BoolFormula.False) => lowRes
+      case (_, _) => boolAlg.mkOr(lowRes, highRes)
+    }
+  }
+
+  class HybridFormula(val formula: Formula, val vars: SortedSet[Int]) extends Formula {
+    val size = vars.size
+  }
+
+  implicit val AsBoolAlg: BoolAlg[HybridFormula] = new BoolAlg[HybridFormula] {
+    override def isTrue(f: HybridFormula): Boolean = f.formula match {
+      case boolForm: BoolFormula => boolAlg.isTrue(boolForm)
+      case bddForm: BddFormula => bddAlg.isTrue(bddForm)
+    }
+
+    override def isFalse(f: HybridFormula): Boolean = f.formula match {
+      case boolForm: BoolFormula => boolAlg.isFalse(boolForm)
+      case bddForm: BddFormula => bddAlg.isFalse(bddForm)
+    }
+
+    override def isVar(f: HybridFormula): Boolean = f.formula match {
+      case boolForm: BoolFormula => boolAlg.isVar(boolForm)
+      case bddForm: BddFormula => bddAlg.isVar(bddForm)
+    }
+
+    override def mkTrue: HybridFormula = new HybridFormula(boolAlg.mkTrue, SortedSet.empty)
+
+    override def mkFalse: HybridFormula = new HybridFormula(boolAlg.mkFalse, SortedSet.empty)
+
+    override def mkVar(id: Int): HybridFormula = new HybridFormula(boolAlg.mkVar(id), SortedSet(id))
+
+    override def mkNot(f: HybridFormula): HybridFormula = f.formula match {
+      case boolForm: BoolFormula => new HybridFormula(boolAlg.mkNot(boolForm), f.vars)
+      case bddForm: BddFormula => new HybridFormula(bddAlg.mkNot(bddForm), f.vars)
+    }
+
+    override def mkOr(f1: HybridFormula, f2: HybridFormula): HybridFormula = {
```

Comment:
Looking at the code, I am wondering if this is "necessary"?

Reply:
What I mean is; I was expecting something like the "strategy" pattern:

- We get an SVE query. If vars is small then use BoolFormulaAlg otherwise use BddFormulaAlg

Maybe we don't even need HybridFormula and a simple if inside BoolUnf is enough?

Reply:
Does it make sense what I am saying?

Reply:
Yes this does make sense. I thought of it more like a State pattern, where we switch back and forth between the representations as the formulas grow and shrink. That is what I have implemented.
If you instead want what you are suggesting, then that is a different solution.
Should I try to make what you are suggesting?

Reply:
Yes, I think so.

(I also think I was not being clear and that I did not know myself what I wanted :))

Reply:
(One this is done, we should redo the experiments.)

Reply:
So I need a line like this:
`implicit val alg: BoolAlg[Formula] =
      if (typeVars.size > varThreshold) BoolFormula.AsBoolAlg
      else BddFormula.AsBoolAlg`
in BoolUnification, but the trouble is that BoolAlg[F] is invariant in F. How can I declare alg such that both variants are allowed? (Both BoolFormula and BddFormula are subtypes of Formula)

Reply:
Ah yes. Uhm. I think you have to have a bigger if:

```
if (...) {
 // do everything with boolAlgFormula and call SVE etc.
} else {
  // do everything with boolgAlgBdd and call SVE etc.
}
```

I think you also need to introduce two caches (one for each type).

Reply:
Does that help?

Reply:
Yes that should work without any problems. Although it is pretty ugly with a lot of code duplication...
I'll have some experiment results in a bit

Reply:
Here are the new results:
| Formula type | min TP (l/s) | max TP | avg TP | median TP | time (s) |
| ------------ | ------------ | ------ | ------ | --------- | -------- |
| BoolFormula  | 5.759        | 19.690 | 16.538 | 18.133    | 43       |
| BddFormula   | 3.111        | 5.098  | 4.756  | 5.007     | 134      |
| Hybrid (> 4) | 5.918        | 19.817 | 17.018 | 19.501    | 41       |
| Hybrid (> 3) | 5.666        | 19.443 | 16.359 | 17.964    | 43       |
| Hybrid (> 2) | 3.090        | 5.168  | 4.761  | 4.944     | 134      |

Reply:
Sweeeet!

Reply:
CC @athas @jacopol

I think this is a very nice preliminary result. The question now is if we can optimize the BDD->Explicit Formula translation such that BDDs are competitive for 3 variables.

Reply:
That looks pretty good, and I think further optimisations are going to make BDDs viable for the >3 case.  That means you can significantly shrink the formula table.  (Not a big deal, but not having a 4MiB table in memory is always a win.)